module/zfs/dbuf.c

```diff
@@ -2784,15 +2789,45 @@ __dbuf_hold_impl(struct dbuf_hold_impl_data *dh)
 	    dh->dh_dn->dn_object != DMU_META_DNODE_OBJECT &&
 	    dh->dh_db->db_state == DB_CACHED && dh->dh_db->db_data_pending) {
 		dh->dh_dr = dh->dh_db->db_data_pending;
-
 		if (dh->dh_dr->dt.dl.dr_data == dh->dh_db->db_buf) {
```

Comment:
For the sake of readability what do you think about moving the contents of this conditional in to a helper function.  We can declare it `noinline static void` to prevent gcc from inline it and increasing the stack depth.  This should also let you avoid having to add additional member to the `dbuf_hold_impl_data` struct.

Reply:
Which conditional are you referring to?  In line 2792, both ```dh_dr``` and ```dh_db``` were already in the ```dbuf_hold_impl_data``` struct.  I'm not sure how this could be made more readable versus the rest of the function.

Reply:
@dweeezil sorry I wasn't very clear.  What I'm suggesting is pulling this entire block out in to a helper function like this, behlendorf/zfs@950e2b1011725978aa8a1ffaed831396a63d191c (untested).  As long as the function is marked `noinline` we should be able to prevent gcc from inlining it and contributing to the stack depth.  Then we can use some local variable to make it more readable.  What do you think?

Reply:
@behlendorf OK, that makes sense.  I'll refactor it as you suggest and rerun the tests.  For future regression testing,  maybe we could stash the send stream I'm using (50MiB) in the zfs-images repo or maybe even make a new repo for send streams.

Reply:
Sure, that would make sense.  Another idea which was floated was to provide a counterpart to `zstreamdump` which would make it easy to generate interesting streams for testing.